File: plsql-checks/src/main/java/br/com/felipezorzo/sonar/plsql/checks/CollapsibleIfStatementsCheck.java
package br.com.felipezorzo.sonar.plsql.checks;

import java.util.List;

import org.sonar.api.server.rule.RulesDefinition;
import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.squidbridge.annotations.ActivatedByDefault;
import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
import org.sonar.sslr.ast.AstSelect;

import com.sonar.sslr.api.AstNode;

import br.com.felipezorzo.sonar.plsql.api.PlSqlGrammar;
import br.com.felipezorzo.sonar.plsql.api.PlSqlKeyword;

@Rule(
    key = CollapsibleIfStatementsCheck.CHECK_KEY,
    priority = Priority.MAJOR,
    tags = Tags.CLUMSY
)
@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.READABILITY)
@SqaleConstantRemediation("5min")
@ActivatedByDefault
public class CollapsibleIfStatementsCheck extends BaseCheck {

    public static final String CHECK_KEY = "CollapsibleIfStatements";

    @Override
    public void init() {
        subscribeTo(PlSqlGrammar.IF_STATEMENT);
    }

    @Override
    public void visitNode(AstNode node) {
        AstNode singleIfChild = singleIfChild(node);
        if (singleIfChild != null && !hasElseOrElif(singleIfChild)) {
            getContext().createLineViolation(this, getLocalizedMessage(CHECK_KEY), singleIfChild);
        }
    }

    private boolean hasElseOrElif(AstNode ifNode) {
        return ifNode.hasDirectChildren(PlSqlKeyword.ELSIF) || ifNode.hasDirectChildren(PlSqlKeyword.ELSE);
    }

    private AstNode singleIfChild(AstNode suite) {
        List<AstNode> statements = suite.getChildren(PlSqlGrammar.STATEMENT);
        if (statements.size() != 1) return null;
        
        AstSelect nestedIf = statements.get(0).select().children(PlSqlGrammar.IF_STATEMENT);
        if (nestedIf.size() != 1) return null;
            
        return nestedIf.get(0);
    }

}

File: zpa-checks/src/main/kotlin/com/felipebz/zpa/checks/CollapsibleIfStatementsCheck.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2026 Felipe Zorzo
 * mailto:felipe AT felipezorzo DOT com DOT br
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.felipebz.zpa.checks

import com.felipebz.flr.api.AstNode
import com.felipebz.zpa.api.PlSqlGrammar
import com.felipebz.zpa.api.annotations.*
import com.felipebz.zpa.api.syntax.IfStatement
import com.felipebz.zpa.api.syntax.SyntaxViews

@Rule(priority = Priority.MAJOR, tags = [Tags.CLUMSY])
@ConstantRemediation("5min")
@RuleInfo(scope = RuleInfo.Scope.ALL)
@ActivatedByDefault
@OptIn(ZpaExperimentalApi::class)
class CollapsibleIfStatementsCheck : AbstractBaseCheck() {

    override fun init() {
        subscribeTo(SyntaxViews.IF_STATEMENT, ::visitIfStatement)
    }

    private fun visitIfStatement(statement: IfStatement) {
        val nestedIf = singleIfChild(statement)
        if (nestedIf != null && !hasElseOrElsif(statement) &&
                !nestedIf.hasDirectChildren(PlSqlGrammar.ELSIF_CLAUSE, PlSqlGrammar.ELSE_CLAUSE)) {
            addIssue(nestedIf, getLocalizedMessage())
        }
    }

    private fun hasElseOrElsif(statement: IfStatement): Boolean {
        return statement.elsifBranches.isNotEmpty() || statement.elseBranch != null
    }

    private fun singleIfChild(statement: IfStatement): AstNode? {
        return statement.statementAstNodes.singleOrNull()
            ?.getFirstChildOrNull(PlSqlGrammar.IF_STATEMENT)
    }

}
